File: src/day02/solution.py
import typing as t


def remap(cpass: str):
    policy, passwd = cpass.split(": ")
    rangenum, code = policy.split(" ")
    s, e = rangenum.split("-")
    return {
        "code": code,
        "s": int(s),
        "e": int(e),
        "pass": passwd
    }


def solve_a(input_mapped: t.List[dict]) -> int:
    valid = 0
    for data in input_mapped:
        total_char = data["pass"].count(data["code"])
        if total_char >= data["s"] and total_char <= data["e"]:
            valid += 1
    return valid


def solve_b(input_mapped: t.List[dict]) -> int:
    valid = 0
    for data in input_mapped:
        fn_val = data["pass"][data["s"] - 1] == data["code"]
        ln_val = data["pass"][data["e"] - 1] == data["code"]
        if fn_val:
            if not ln_val:
                valid += 1
        elif ln_val:
            if not fn_val:
                valid += 1
    return valid


if __name__ == "__main__":
    in_data = [num.rstrip() for num in open("input", "r").readlines() if num]
    # in_data = ["1-3 a: abcde", "1-3 b: cdefg", "2-9 c: ccccccccc"]
    mapped_data = [remap(c) for c in in_data]
    print(f"Part A: {solve_a(mapped_data)}")
    print(f"Part B: {solve_b(mapped_data)}")

File: src/day02/solution.js
const utils = require("../utils/utils");

let in_data = utils.readInput("input", true);

let remapped_data = [];
for (let index = 0; index < in_data.length; index++) {
    let element = in_data[index];
    let [policy, passwd] = element.split(": ");
    let [rangenum, code] = policy.split(" ");
    let [start, end] = rangenum.split("-");
    remapped_data.push({
        code: code,
        pass: passwd.trimRight(),
        s: parseInt(start),
        e: parseInt(end)
    });
}

// Part A
let valid_a = 0;
let valid_b = 0;
remapped_data.forEach((data) => {
    let total_char = data.pass.length;
    let total_char_rep = data.pass.replace(new RegExp(utils.escapeRegExp(data.code), "g"), "").length;
    let diff = total_char - total_char_rep;
    let fn_val = data.pass.charAt(data.s - 1) == data.code;
    let ln_val = data.pass.charAt(data.e - 1) == data.code;
    if (diff >= data.s && diff <= data.e) {
        valid_a++;
    }
    if (fn_val && !ln_val) {
        valid_b++;
    } else if (ln_val && !fn_val) {
        valid_b++;
    }
})

console.log(`Part A: ${valid_a}`);
console.log(`Part B: ${valid_b}`);
